components/Comment/lib/xrange/XPathUtil.ts

```diff
@@ -0,0 +1,221 @@
+function XPathUtil() {}
+
+// Example: "/html/body/div/p[2]" will return "/html/body/div/p[1]"
+// If not matching sibling is possible, will return null
+// Does not validate against the DOM that the sibling is valid
+XPathUtil.findPrevMatchingSibling = (xpathStr) => {
```

Comment:
man.. there's a lot going on here. Could you briefly explain what's happening here

Reply:
Yeah, I realize that. I imported it from another project I owned. I think rather than explaining each function, I will annotate each file because there are a lot of helper functions in here and it may be overkill to try and explain each one.